reana_server/utils.py

```diff
@@ -136,6 +138,68 @@ def mv_workflow_files(source: str, target: str) -> None:
         shutil.move(os.path.join(source, entry), target)
 
 
+# FIXME: use `is_relative_to` from the standard library when moving to Python 3.9
+def is_relative_to(path: pathlib.Path, base: pathlib.Path) -> bool:
+    """Check whether `path` is contained inside `base`."""
+    try:
+        path.relative_to(base)
+        return True
+    except ValueError:
+        return False
+
+
+def filter_input_files(workspace: Union[str, pathlib.Path], reana_spec: Dict) -> None:
+    """Delete the files and directories not specified as inputs in the specification.
+
+    :param workspace: Path to the directory containing the files to be filtered.
+    :param reana_spec: REANA specification used to decide which files to keep.
+    """
+    inputs = reana_spec.get("inputs", {})
+    files = inputs.get("files", [])
+    directories = inputs.get("directories", [])
+
+    if isinstance(workspace, str):
+        workspace = pathlib.Path(workspace)
+
+    for file in files:
+        file_path = workspace / file
+        if not file_path.exists():
+            raise REANAValidationError(f"Input file not found: {file}")
+
+    for directory in directories:
+        directory_path = workspace / directory
+        if not directory_path.exists():
+            raise REANAValidationError(f"Input directory not found: {directory}")
+
+    paths = [pathlib.Path(path) for path in files + directories]
+    filtered = workspace / f"filtered-{uuid4()}"
+    filtered.mkdir()
+
+    # Move input files to the temporary `filtered` directory
+    for path in paths:
+        full_source_path = workspace / path
+        full_target_path = filtered / path
+        # Create target directory if it does not exist
+        full_target_path.parent.mkdir(parents=True, exist_ok=True)
+        full_source_path.replace(full_target_path)
+
+    # Delete remaining files in the workspace
+    for path in workspace.iterdir():
+        if path == filtered:
+            continue
+        if path.is_file() or path.is_symlink():
+            path.unlink()
+        else:
+            shutil.rmtree(path)
+
+    # Move the contents of the temporary `filtered` directory to the workspace
+    for path in filtered.iterdir():
+        path.replace(workspace / path.name)
+
+    # Remove temporary `filtered` directory
+    filtered.rmdir()
```

Comment:
*note(non-blocking):* Initially, I was thinking maybe it is better to just delete files/directories that are not in `input` instead of moving them to `filtered-<uuid>` and back. I came up with this:

```python
    for path in workspace.iterdir():
        if any([is_relative_to(workspace / input_path, path) for input_path in paths]):
            continue
        if path.is_file() or path.is_symlink():
            path.unlink()
        else:
            shutil.rmtree(path)
```

But my code has an issue, it can copy unnecessary files. I think it is possible to do it with `os.walk` but I am not sure it will give any performance benefits. Move operations are pretty fast. Anyway, leaving it here as a learning/observation.